[WIP] Conversation API Implementation

parlough opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1532) | **SpongeCommon** | [SpongeVanilla](https://github.com/SpongePowered/SpongeVanilla/pull/310) | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1427)

The work in progress SpongeCommon implementation of my Conversation API.

parlough committed:
Functional common conversation implementation

Meronat commented:
I cannot figure out what I did wrong with the submodules for this and the Vanilla and Forge PRs which seems to be causing the build to fail. Could someone provide some assistance there?

kashike commented on src/main/java/org/spongepowered/common/conversation/SpongeConversationManager.java:
```diff
@@ -0,0 +1,192 @@
+
+    @Override
+    public Collection<Conversation> getConversation(PluginContainer plugin, String id) {
+        return conversations.get(plugin.getId() + ":" + id.toLowerCase());
```
`this.`

kashike commented on src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java:
```diff
@@ -404,10 +407,21 @@ public boolean isChatColorsEnabled() {
+            return;
+        }
+
         ITextComponent component = SpongeTexts.toComponent(message);
```
You should move this logic to a method like `sendMessage0` and call it from `sendMessage` and `sendThroughMessage`.

parlough replied:
So I don't want this logic within `sendThroughMessage` actually, but I would like some input.

The `sendThroughMessage` method was created, because you can set a conversation to catch `Conversant` input, which would include the messages you are trying to send to them while asking questions and similar. I couldn't think of the greatest way to accomplish besides making a new method which bypasses this catching. 

I would be happy to have some input on this.

kashike commented on src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java:
```diff
@@ -404,10 +407,21 @@ public boolean isChatColorsEnabled() {
     @Override
+    public void sendThroughMessage(Text message) {
+        checkNotNull(message, "The message text cannot be null!");
+        this.connection.sendPacket(new SPacketChat(SpongeTexts.toComponent(message)));
```
See message regarding `sendMessage`.

kashike commented on src/main/java/org/spongepowered/common/event/conversation/SpongeConversationEndCause.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public String toString() {
+        return Objects.toStringHelper("ConversationEndCause")
```
`toStringHelper(this)`

parlough committed:
Small style and convention fixes, and tries to fix submodule problem

stephan-gh commented on src/main/java/org/spongepowered/common/conversation/SpongeConversation.java:
```diff
@@ -0,0 +1,219 @@
+        cause = cause.with(NamedCause.of(endType.getName(), ConversationEndCause.builder().type(endType).build()));
+        final ConversationCloseEvent.Ending endingEvent = SpongeEventFactory.createConversationCloseEventEnding(cause, this);
+
+        if (SpongeImpl.postEvent(endingEvent) || endingEvent.isCancelled()) {
```
`SpongeImpl.postEvent` returns whether the event was (not?) cancelled so the extra check isn't necessary.

stephan-gh commented on src/main/java/org/spongepowered/common/conversation/SpongeConversationManager.java:
```diff
@@ -0,0 +1,192 @@
+                        conversation.setQuestion(question);
+                        SpongeImpl.getLogger().error("Failed to parse the arguments passed into " + conversant.getName() +
+                            "'s conversation." + " Sending them the same question again.");
+                        e.printStackTrace();
```
You should give the exception to the logger message above and use placeholders instead of building the message manually. Basically like:

```java
SpongeImpl.getLogger().error("Failed to parse the arguments passed into {}'s conversation. Sending them the same question again.", conversant.getName(), e);
```

stephan-gh commented on src/main/java/org/spongepowered/common/conversation/SpongeConversationManager.java:
```diff
@@ -0,0 +1,192 @@
+                                break;
+                            default:
+                                // Should not occur
+                                SpongeImpl.getLogger().error("For some reason the question result type did not match any.");
```
`this.logger`

stephan-gh commented on src/main/java/org/spongepowered/common/conversation/SpongeQuestion.java:
```diff
@@ -0,0 +1,77 @@
+        this.arguments = arguments;
+    }
+
+    @Nonnull
```
This should be already handled by the `@NonnullByDefault` annotation in the package.

parlough replied:
Yeah I noticed I left quite a lot of these around, I'll remove them next commit.

kashike commented on src/main/java/org/spongepowered/common/conversation/SpongeConversationArchetype.java:
```diff
@@ -0,0 +1,166 @@
+        this.exit = exit.toLowerCase();
+        this.title = title;
+        this.padding = padding;
+        if (title != null && padding != null) {
```
Header should be controllable by the plugin.

parlough replied:
A header is generally composed of a title and it's padding like pagination but I like this idea. If a header is set then that would override those so it could be whatever they so desire.

parlough replied:
Or do you think the header should be something separate from the title and padding such as in pagination?

parlough committed:
Try to let Sponge handle submodule

parlough committed:
Fix some logging style issues

parlough committed:
Not sure how these imports didn't update

parlough committed:
Add new end type, fix some style issues, and try to account for loss of conversant references

stephan-gh commented on src/main/java/org/spongepowered/common/event/conversation/SpongeConversationEndCauseBuilder.java:
```diff
@@ -0,0 +1,44 @@
+public class SpongeConversationEndCauseBuilder extends AbstractConversationEndCauseBuilder<ConversationEndCause, Builder>
+    implements ConversationEndCause.Builder {
+
+    @Nonnull
```
Unneeded, should be handled by `@NonnullByDefault`

stephan-gh commented on src/main/java/org/spongepowered/common/text/conversation/SpongeConversation.java:
```diff
@@ -0,0 +1,256 @@
+                .interval(5, TimeUnit.MINUTES)
+                .execute(() -> {
+                    if (this.externalChatHandlers.size() == 0) {
+                        SpongeImpl.getLogger().error("A " + creator.getId() + ":" + getId().toLowerCase() + " conversation has ended due to being "
```
Add a locale for the call to `toLowerCase()` (usually `Locale.ROOT` is appropriate for these)

parlough commented:
To-do for myself:

- [x] Fix compilation errors

- [x] Switch to Guava's MapMaker instead of my stupid Synchronized map usage

- [x] Add tab completion support

- [ ] Add console as a conversant

- [x] Allow optional header setting

- [x] Remove unnecessary annotations

- [x] Allow setting input tokenizer on question builder

- [x] Allow changing the no command usage message

- [x] Move ExternalChatHandlers to some sort of catalog type

parlough committed:
Fix some compile issues and move to command package

parlough committed:
Movement to contextual prompts

parlough committed:
Add support for setting a header

kashike commented on src/main/java/org/spongepowered/common/command/SpongeCommandManager.java:
```diff
@@ -266,6 +266,16 @@ public boolean containsMapping(CommandMapping mapping) {
+            if (optionalConversation.isPresent()) {
+                final Conversation conversation = optionalConversation.get();
+                if (!conversation.allowsCommands()) {
+                    ((Conversant) source).sendThroughMessage(Text.of("You must exit this conversation before you can use commands!"));
```
I'd prefer if the conversation/plugin can handle this, to allow this message to be customised.

parlough replied:
Will do :)

parlough committed:
Adds tab completion functionality

To accomplish this I refactored some of the command manager's tab completion and removed our getTabCompletion mixins, replacing it with a complete overwrite since we change so much

parlough committed:
Allow command usage message and input tokenizer to be set.

Also cleanup the builders in general and add a few comments

parlough committed:
Add cancelling handlers

parlough committed:
Don't need to implement this method

kashike commented on src/main/java/org/spongepowered/common/command/conversation/SpongeQuestion.java:
```diff
@@ -0,0 +1,87 @@
+import org.spongepowered.api.command.conversation.AnswerHandler;
+import org.spongepowered.api.command.conversation.PromptHandler;
+import org.spongepowered.api.command.conversation.Question;
+public class SpongeQuestion implements Question {
```
Missing new line

parlough committed:
Begin work on supporting console sources

parlough committed:
Some small fixes

parlough committed:
Continue new external chat handler system

parlough commented:
If anyone could provide some insight, my latest commit does not build due to the following log: https://gist.github.com/Meronat/056ed67062bdf549610843b7d977a0e0

Seems to be an issue with Guice and my SpongeConversationManager, mainly around this piece of the log:

```
1) Error injecting constructor, java.lang.IllegalArgumentException: The provided plugin object does not have an associated plugin container (in other words, is 'plugin' actually your plugin object?
      at org.spongepowered.common.command.conversation.SpongeConversationManager.<init>(SpongeConversationManager.java:86)
      at org.spongepowered.common.command.conversation.SpongeConversationManager.class(SpongeConversationManager.java:86)
      while locating org.spongepowered.common.command.conversation.SpongeConversationManager
      while locating org.spongepowered.api.command.conversation.ConversationManager
        for field at org.spongepowered.api.Sponge.conversationManager(Sponge.java:52)
```

This does not really make sense to me anyway since there is no plugin being passed into the line there(constructor), simply a logger.

kashike commented:
It sounds as if you're registering a command before the plugin container for the implementation has been created and is available.

parlough commented:
@kashike That makes quite a lot of sense actually. Thanks!  Will fix that soon.

parlough committed:
Fix compile issues with external chat handler catalog type

parlough committed:
Fixes to some suggestions

parlough committed:
Oops

kashike commented on src/main/java/org/spongepowered/common/command/conversation/SpongeConversationManager.java:
```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    public static CommandSpec getCommand() {
```
`createCommand`

parlough committed:
Move question result to implementation with a builder

parlough committed:
Fixes for feedback on PR, still has tuning to do

ImMorpheus commented:
Closing this given the age that has passed and the lack of activity. If someone wants to continue the work done here then we can re-open at a later time.
